server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java

```diff
@@ -295,6 +301,38 @@ protected void search(List<LeafReaderContext> leaves, Weight weight, Collector c
                 searchLeaf(leaves.get(i), weight, collector);
             }
         }
+        postCollection(collector);
+    }
+
+    /**
+     * Performs the postCollection on the {@link BucketCollector} which is used to collect documents for Aggregation. This was originally
+     * done in {@link org.opensearch.search.aggregations.AggregationProcessor#postProcess(SearchContext)}. But with concurrent segment
+     * search path this needs to be performed here. There are AssertingCodecs in lucene which validates that the DocValues created for a
+     * field is always used by the same thread for a request. In concurrent segment search case, the DocValues gets initialized on
+     * different threads for different segments (or slices). Whereas the postProcess happens as part of reduce phase and is performed on
+     * the separate thread which is from search threadpool and not from slice threadpool. So two different threads performs the access on
+     * the DocValues causing the AssertingCodec to fail. From functionality perspective, there is no issue as DocValues for each segment
+     * is always accessed by a single thread at a time but those threads may be different (e.g. slice thread during collection and then
+     * search thread during reduce)
+     * @param collector input collector tree used for search
+     */
+    private void postCollection(Collector collector) throws IOException {
+        final Queue<Collector> collectors = new LinkedList<>();
+        collectors.offer(collector);
```

Comment:
@sohami this is super fragile, I think we should model it using something like: 
```
public interface FinalizingCollector {
    void postCollection() throws IOException;
}
```

The `BucketCollector` naturally implements that but for InternalProfileCollector / MultiCollector / MinimumScoreCollector / ... others we should implement that by checking the wrapping collector(s).

Reply:
@reta thanks for sharing the link. I was also planning to create a separate issue in lucene to provide some `postCollect` option in `Collector` interface itself. Considering that will take time to implement and merge this was a workaround meanwhile. But thanks for sharing the link seems like it is getting addressed already, however until that get release we still need other mechanism.

I had thought about The `FinalizingCollector` interface mechanism but that is same as providing the capability in `Collector` interface itself. This mechanism will mean for collectors like `MultiCollector` and others we need to create a wrapper in OpenSearch and use that wrapper everywhere to make it work. Also I was not sure if that `MultiCollectorWrapper` will also get wrapped inside the vanilla `MultiCollector` by lucene in some code path then this interface mechanism will not work. In current mechanism, for `MultiCollector` handling we are unwrapping it recursively until a concrete `Collector` is found, so that issue will not be there. Given the lucene is changing the interface to add the method in `LeafCollector`, I think we can keep this current implementation as a stop gap solution with a follow up issue to move to lucene `postCollect` mechanism once that is merged in. Let me know wdyt ?

Reply:
Oh right, I though `MultiCollector` is ours, but it is from Apache Lucene indeed :( However, we could (and probably should move this out of `ContextIndexSearcher`:
 - this is very relevant to aggregations only and not the regular search flow
 - we already have mostly identical code in AggregationCollectorManager::reduce

What about we introduce the `SearchContext::postProcessor()`  (similarly to what we did with `ReduceContext` and the postrocessor will only be set during the aggregation flows (and also implemented within aggregation framework), otherwise - NO_OP

Reply:
We want this `postCollection` to be done by individual slice threads for the collector instance it is using and thats why it is kept here as each slice related search operation is performed here. This way we don't need to worry about the thread safety as each slice thread will work on its own collector instance.

> this is very relevant to aggregations only and not the regular search flow

We can add a check like `searchContext::aggregations() != null` here before invoking this method.  This will take care of not performing it when aggregations are not present.

> we already have mostly identical code in AggregationCollectorManager::reduce

There is difference between unwrapping of collectors here and in `AggregationCollectorManager::reduce`. In later, the `CollectorManager` expects the input collectors to be the one created by the manager and not the other collectors in the tree. But that is not the case here.

Reply:
> This way we don't need to worry about the thread safety as each slice thread will work on its own collector instance.

Fair enough, let's go with `SearchContext::newPostProcessor()` (or alike) to eliminate any ambiguity here. I really feel uneasy about this code in `ContextIndexSearcher` there is very far from were it should be (aggregations).

Reply:
Do you mean the check for `searchContext::aggregations() != null` right ? Or something else here. I am trying to avoid too much changes here as this is temporary and we want to remove this once the lucene changes are merged in.

Reply:
> searchContext::aggregations() != null

No, the `ContextIndexSearcher` should be free of this knowledge, I mean to replace

```
postCollection(collector)
```

with (pseudo code):

```
searchContext.newPostProcessor().postProcess(collector);
```

The `searchContext` will be configured with post processors only within aggregation processor(s)

Reply:
> No, the ContextIndexSearcher should be free of this knowledge,

May be I am missing something but I don't think that is the case with `ContextIndexSearcher`, it already takes action based on the request context today. It has access to `SearchContext` and we do optimizations like for [sort](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L510) and [dfs-phase](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L199) in this class. So I don't see why we cannot perform this operation for `aggregation` here. Today, the `postCollect` is implemented by `Aggregations` only but if we had it in generic collector interface then it should be called for each of the collectors and that is what now lucene is moving towards. The processing here is done on `Collector` and by different slice threads and IMO `SearchContext` is not meant to provide any capability to perform the operations across threads and is used for keeping track of request level attributes.

Reply:
> t has access to `SearchContext` and we do optimizations like for [sort](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L510) and [dfs-phase](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L199) in this class.

I am suggesting to use `searchContext` for that (see please previous comment). I would argue that index searcher should know  nothing:
  a)  about aggregation 
  b)  about different types of collectors

The fact it has some more knowledge now is not a reason to do the same

>  if we had it in generic collector interface then it should be called for each of the collectors and that is what now lucene is moving towards. 

The issue I posted to you - please feel free to contribute (either by commenting or submitting the pull request), it may take a long time otherwise (I have no insight)

> The processing here is done on Collector and by different slice threads and IMO SearchContext is not meant to provide any capability to perform the operations across threads

I 100% agree with you here, where you see the mentioning of threads or across threads or anything related to threads in the suggestion to move the implementation off the `ContextIndexSearcher`? Is it very easy to copy / replicate same code in different places but it is very difficult to support that. As I mentioned several times, we sadly have to make some adjustments in the post processing phase - but this code should live in aggregations related implementation. If you see any other ways to achieve that - please share.

Reply:
@reta Sorry for the delay, have made the changes to create a new processor to keep the logic for collector processing and which gets initialized in Aggregation flow in `SearchContext`.